pyiceberg/table/__init__.py

```diff
@@ -868,6 +919,12 @@ def sort_orders(self) -> Dict[int, SortOrder]:
         """Return a dict of the sort orders of this table."""
         return {sort_order.order_id: sort_order for sort_order in self.metadata.sort_orders}
 
+    def last_partition_id(self) -> Optional[int]:
+        """Return the highest assigned partition field ID across all specs for the table or None if the table is unpartitioned and there are no specs."""
+        if len(self.specs()) == 1 and self.spec().is_unpartitioned():
+            return None
+        return self.metadata.last_partition_id
```

Comment:
@Fokko @HonahX I added this API to the `Table` since we'll need it in the implementation and we don't want to directly access `TableMetadata` let me know what you think

Reply:
I think we probably should update https://github.com/apache/iceberg-python/blob/7fbcc220263228618da8d6871a50a0b82e22a843/pyiceberg/partitioning.py#L148-L152

to return PARTITION_FIELD_ID_START - 1 for unpartitioned spec. Then we can return the last_partition_id from metadata directly because the metadata should have `last_partition_id=999` for unpartitioned table.

Java implementation uses PARTITION_FIELD_ID_START - 1 for unpartitioned spec:
https://github.com/apache/iceberg/blob/main/api/src/main/java/org/apache/iceberg/PartitionSpec.java#L344-L345
https://github.com/apache/iceberg/blob/9921937d8285dec9a19fd16b0cd82d451a8aca9e/api/src/main/java/org/apache/iceberg/PartitionSpec.java#L319-L321

I checked locally that unpartitioned tables created by spark-iceberg-runtime have `last_partition_id=999`, while those created by pyiceberg have `last_partition_id=1000`:

Reply:
I was thinking about that but wasn't sure about breaking API compatibility which is why I added a new API. If we have the flexibility here to change the API we should do that. I think we can because arguably it's incorrect to return 1000 for an unpartitioned table so it's really a fix.